### PR TITLE
Add the Hotplug agent

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by nomis52** [comment]
Blocked behind #925

File: common/thread/PeriodicThread.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  while (true) {
+    {
+      MutexLocker lock(&m_mutex);
+      if (m_terminate) {
+        return NULL;
+      }
+      if (m_condition.TimedWait(&m_mutex, last_run_at + m_delay)) {
+        // Either m_terminate is true, or a suprious wake up
```
Comment (peternewman) [inline common/thread/PeriodicThread.cpp]: SPaG spurious

File: common/thread/PeriodicThread.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  Join(NULL);
+}
+
+void *PeriodicThread::Run() {
+  Clock clock;
+  TimeStamp last_run_at;
+
+  clock.CurrentTime(&last_run_at);
+  if (!m_callback->Run()) {
+    return NULL;
```
Comment (peternewman) [inline common/thread/PeriodicThread.cpp]: Just return, given it's void?

File: common/thread/PeriodicThread.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  while (true) {
+    {
+      MutexLocker lock(&m_mutex);
+      if (m_terminate) {
+        return NULL;
+      }
+      if (m_condition.TimedWait(&m_mutex, last_run_at + m_delay)) {
+        // Either m_terminate is true, or a suprious wake up
+        if (m_terminate) {
+          return NULL;
```
Comment (peternewman) [inline common/thread/PeriodicThread.cpp]: Again

File: common/thread/PeriodicThread.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  clock.CurrentTime(&last_run_at);
+  if (!m_callback->Run()) {
+    return NULL;
+  }
+
+  while (true) {
+    {
+      MutexLocker lock(&m_mutex);
+      if (m_terminate) {
+        return NULL;
```
Comment (peternewman) [inline common/thread/PeriodicThread.cpp]: Again

File: common/thread/PeriodicThread.cpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        // Either m_terminate is true, or a suprious wake up
+        if (m_terminate) {
+          return NULL;
+        }
+        continue;
+      }
+    }
+    clock.CurrentTime(&last_run_at);
+    if (!m_callback->Run()) {
+      return NULL;
```
Comment (peternewman) [inline common/thread/PeriodicThread.cpp]: Again

File: common/thread/PeriodicThread.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    }
+  }
+  return NULL;
```
Comment (peternewman) [inline common/thread/PeriodicThread.cpp]: Again

File: libs/usb/HotplugAgent.h [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  bool m_use_hotplug;
+  libusb_context *m_context;
+  std::auto_ptr<ola::usb::LibUsbThread> m_usb_thread;
+  std::auto_ptr<ola::usb::AsyncronousLibUsbAdaptor> m_usb_adaptor;
+  std::auto_ptr<ola::thread::PeriodicThread> m_scanner_thread;
+
+  ola::thread::Mutex m_mutex;
+  bool m_suppress_hotplug_events;  // GUARDED_BY(m_mutex);
+
+  // In hotplug mode, this is guardded by m_mutex while
```
Comment (peternewman) [inline libs/usb/HotplugAgent.h]: SPaG guarded

File: tools/ja-rule/USBDeviceManager.cpp [hunk 8]
```diff
@@ -37,118 +35,86 @@
...
+    if (iter->second) {
+      m_notification_cb->Run(WIDGET_REMOVED, iter->second);
+      m_cleanup_thread.Execute(ola::DeletePointerCallback(iter->second));
     }
-    m_widgets.clear();
   }
+  m_widgets.clear();
 
-  // Blocks until all devices have been deleted.
+  // Blocks all widgets have been deleted.
```
Comment (peternewman) [inline tools/ja-rule/USBDeviceManager.cpp]: SPag Blocks until

**Comment by peternewman** [comment]
Generally LGTM

**Comment by peternewman** [comment]
Can you also uncomment and re-add the --enable-ja-rule bits in .travis.yml and .travis-ci.sh please, as that will now work as it doesn't require the latest libusb.

File: common/thread/PeriodicThread.cpp (hunk 2 above)
Comment (nomis52) [inline common/thread/PeriodicThread.cpp]: It's void* which is why we return NULL

File: common/thread/PeriodicThread.cpp (hunk 4 above)
Comment (nomis52) [inline common/thread/PeriodicThread.cpp]: Same for all of these.

File: common/thread/PeriodicThread.cpp (hunk 1 above)
Comment (nomis52) [inline common/thread/PeriodicThread.cpp]: Fixed

File: libs/usb/HotplugAgent.h (hunk 7 above)
Comment (nomis52) [inline libs/usb/HotplugAgent.h]: Fixed

File: tools/ja-rule/USBDeviceManager.cpp (hunk 8 above)
Comment (nomis52) [inline tools/ja-rule/USBDeviceManager.cpp]: Fixed

**Comment by peternewman** [comment]
LGTM can you just do .travis.yml too please.

**Comment by nomis52** [comment]
Done

**Comment by peternewman** [comment]
LGTM

**Pull request merged**